test/chunk_info.c

```diff
@@ -2180,17 +2118,312 @@ test_flt_msk_with_skip_compress(hid_t fapl)
+
+    /* Get the number of chunks */
+    if (H5Dget_num_chunks(did, H5S_ALL, &num_chunks) < 0)
+        TEST_ERROR;
+    if (H5Dget_num_chunks(did_ub, H5S_ALL, &num_chunks_ub) < 0)
+        TEST_ERROR;
+
+    if (num_chunks != num_chunks_ub)
+        TEST_ERROR;
+
+    /* Check the chunk information to make sure that the userblock file takes
+     * the block's size into account.
+     */
+    for (hsize_t i = 0; i < num_chunks; i++) {
+        haddr_t addr    = HADDR_UNDEF;
+        haddr_t addr_ub = 0;
+
+        /* H5Dget_chunk_info() */
+        if (H5Dget_chunk_info(did, H5S_ALL, i, NULL, NULL, &addr, NULL) < 0)
+            TEST_ERROR;
+        if (H5Dget_chunk_info(did_ub, H5S_ALL, i, NULL, NULL, &addr_ub, NULL) < 0)
+            TEST_ERROR;
+
+        if (addr + UBLOCK_SIZE != addr_ub)
+            TEST_ERROR;
+
+        addr    = HADDR_UNDEF;
+        addr_ub = 0;
+
+        /* H5Dget_chunk_info_by_coord() */
+        if (H5Dget_chunk_info_by_coord(did, &i, NULL, &addr, NULL) < 0)
+            TEST_ERROR;
+        if (H5Dget_chunk_info_by_coord(did_ub, &i, NULL, &addr_ub, NULL) < 0)
+            TEST_ERROR;
+
+        if (addr + UBLOCK_SIZE != addr_ub)
+            TEST_ERROR;
+    }
+
+    /* Allocate arrays to hold the chunk addresses */
+    if (NULL == (addresses = (haddr_t *)calloc(num_chunks, sizeof(haddr_t))))
+        TEST_ERROR;
+    if (NULL == (addresses_ub = (haddr_t *)calloc(num_chunks, sizeof(haddr_t))))
+        TEST_ERROR;
+
+    od.addresses = addresses;
+    od.i         = 0;
+    od.max       = num_chunks;
+
+    od_ub.addresses = addresses_ub;
+    od_ub.i         = 0;
+    od_ub.max       = num_chunks;
+
+    /* Iterate over the chunks, storing the chunk addresses */
+    if (H5Dchunk_iter(did, H5P_DEFAULT, ublock_iter_cb, &od) < 0)
+        TEST_ERROR;
+    if (H5Dchunk_iter(did_ub, H5P_DEFAULT, ublock_iter_cb, &od_ub) < 0)
+        TEST_ERROR;
+
+    /* Compare the chunk addresses to ensure the userblock file takes the
+     * chunk's size into account.
+     */
+    if (od.i != od_ub.i)
+        TEST_ERROR;
+    for (hsize_t i = 0; i < num_chunks; i++)
+        if (od.addresses[i] + UBLOCK_SIZE != od_ub.addresses[i])
+            TEST_ERROR;
+
+    /* Compare the raw chunk data */
+    if ((fd = HDopen(NO_UBLOCK_FILE_NAME, O_RDONLY)) < 0)
+        TEST_ERROR;
+    if ((fd_ub = HDopen(UBLOCK_FILE_NAME, O_RDONLY)) < 0)
+        TEST_ERROR;
+
+    for (hsize_t i = 0; i < num_chunks; i++) {
+        int data    = -1;
+        int data_ub = -1;
+
+        if (HDlseek(fd, (off_t)(od.addresses[i]), SEEK_SET) < 0)
+            TEST_ERROR;
+        if (HDlseek(fd_ub, (off_t)(od_ub.addresses[i]), SEEK_SET) < 0)
+            TEST_ERROR;
+
+        if (HDread(fd, &data, sizeof(int)) != sizeof(int))
+            TEST_ERROR;
+        if (HDread(fd_ub, &data_ub, sizeof(int)) != sizeof(int))
+            TEST_ERROR;
+
+        if (data != data_ub)
+            TEST_ERROR;
+    }
+
+    HDclose(fd);
+    HDclose(fd_ub);
```

Comment:
Not a big deal, but reset FDs to -1 in case below closes fail and try to do the close again.